### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by larsimmisch**
This should address #58

### CODE REVIEW

**Comment by ossilator** [comment]
what was wrong with the previous PR?

**Comment by larsimmisch** [comment]
I was just too tired to rebase it to the new `main`, so I created a new branch/PR. 🤷‍♂️

**Comment by ossilator** [comment]
i'm confused. you needed to rebase locally anyway, and updating the PR ist just a forced push into the existing branch.

**Comment by ossilator** [comment]
some random observations from a quick scan:
- you should probably rebase and cover the two new functions i added docu for
- you're being inconsistent about the whitespace after colons
- the trailing whitespace removal should be committed separately. given that you have set up your editor to do that, it's probably best if you just touch all files once, so it's out of the way.

how are the hints rendered in html? do the prototypes look just as you provide them, or does it do something clever? i guess the former, as you need to provide the info redundantly ...

File: doc/libalsaaudio.rst [hunk 1]
```diff
@@ -267,37 +267,29 @@ PCM objects have the following methods:
...
 
-.. method:: PCM.info()
-
-   Returns a dictionary with the PCM object's configured parameters.
-
-   Values are retrieved from the ALSA library if they are available;
-   otherwise they represent those stored by pyalsaaudio, and their keys
-   are prefixed with ' (call value) '.
-
    *New in 0.9.1*
```
Comment (RonaldAJ) [inline doc/libalsaaudio.rst:286]: *New in 0.9.1* belonged to PCM.info so should be removed

Comment (RonaldAJ) [reply doc/libalsaaudio.rst:286]: Can't comment on the technical changes. I would have to read up on type hints before I can.

File: alsaaudio.pyi [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+VOLUME_UNITS_PERCENTAGE: int
+VOLUME_UNITS_RAW: int
+VOLUME_UNITS_DB: int
+
+def pcms(pcmtype:int) -> list[str]: ...
+def cards() -> list[str]: ...
+def mixers(cardindex:int=-1, device:str='default') -> list[str]: ...
+def asoundlib_version() -> str: ...
+
+class PCM:
```
Comment (ossilator) [inline alsaaudio.pyi:83]: some functions are definitely still missing here.

File: doc/libalsaaudio.rst [hunk 3]
```diff
@@ -410,22 +410,15 @@ PCM objects have the following methods:
 
    *New in 0.10*
 
-.. method:: PCM.close()
```
Comment (ossilator) [inline doc/libalsaaudio.rst:413]: if this is legit (is it?), then it definitely belongs into a separate commit.

File: doc/conf.py [hunk 4]
```diff
@@ -35,7 +35,6 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
 extensions = []
-
```
Comment (ossilator) [inline doc/conf.py:38]: how did that get past the whitespace cleanup?

**Review by ossilator — CHANGES_REQUESTED** [review]
the commit message of the second commit is sort of backwards - you're adding type hints to the docu, not the other way around.
but i'd actually squash the first two commits.

the last commit could use a description in the commit message. one sentence should be enough.

note that the comment about whitespace inconsistency still applies.

File: doc/libalsaaudio.rst (hunk 3 above)
Comment (larsimmisch) [reply doc/libalsaaudio.rst:413]: It's legit. `close` wasn't deleted, but moved to the end. Creating a separate commit for that seems overkill to me, though.

File: alsaaudio.pyi (hunk 2 above)
Comment (larsimmisch) [reply alsaaudio.pyi:83]: I didn't document the ones that return `None` and have no arguments, is that it? 

I think type hints are great, but as they're optional, less might be more. In any case, I'll try to double check for consistency

**Comment by larsimmisch** [comment]
> but i'd actually squash the first two commits.

Yeah, clearer, done.

File: doc/libalsaaudio.rst (hunk 3 above)
Comment (ossilator) [reply doc/libalsaaudio.rst:413]: > It's legit. close wasn't deleted, but moved to the end.

yes, i've seen that later.

> Creating a separate commit for that in retrospect seems overkill to me, though.

but this is exactly what good commit hygiene is about. it would have avoided me asking that "stupid" question.

and fixing it would take about 30 seconds in git gui (plus however much you need to type the commit message).

File: alsaaudio.pyi (hunk 2 above)
Comment (ossilator) [reply alsaaudio.pyi:83]: i didn't check, i assume you're right.
but the argument doesn't convince me. firstly, that inconsistency is confusing - the question whether this was omitted by accident will come up each time someone notices it. secondly, i think that file may be used for auto-completion by some IDEs, so having complete declarations is helpful regardless.

**Comment by ossilator** [comment]
totally predictable comment from me: the .pyi change should be squashed onto the first commit.

as for the .c change, i'd be in favor of just dropping it - it's churn that makes following the history (blame) a lot harder, for very little benefit, imo.

if you insist on keeping it, then you should factor it out into a commit that goes at the beginning of the series, so it's in front of the addition that replicates the same order.

what does the "fix todo" refer to?

**Comment by larsimmisch** [comment]
Hi @ossilator,

I've addressed your feedback (thanks) and did another force-push.

What I haven't done yet is validating this change, i.e. *see* the type hints in a native Linux development environment. Maybe I'll install a recent python-mode for Emacs.

> totally predictable comment from me: the .pyi change should be squashed onto the first commit.
:-)

> as for the .c change, i'd be in favor of just dropping it - it's churn that makes following the history (blame) a lot harder, for very little benefit, imo.

I get your point, but I'd rather have `getchannels/setchannels` together.

> if you insist on keeping it, then you should factor it out into a commit that goes at the beginning of the series, so it's in front of the addition that replicates the same order.

Done.

**Comment by ossilator** [comment]
structure-wise it appears correct now.

you're still a bit inconsistent with the whitespace in the type hints. try to standardize on one style, preferably no space before the colon and one space after the colon. though there is a challenge with default params then, because "foo: int=5" looks kinda stupid. maybe surround the equal sign with spaces as well? did the python community develop a canonical style for this?

**Comment by larsimmisch** [comment]
@ossilator I've made the whitespace consistent like you suggested (thanks) and squashed the change into the second commit.

Maybe more importantly, I verified that the type hints are shown using VSCode/Pylance. I noticed that IDE support isn't great without the docstrings, because parameters aren't shown. I'll try to experiment with Sphinx autodoc, but that would be another PR.

> though there is a challenge with default params then, because "foo: int=5" looks kinda stupid. maybe surround the equal sign with spaces as well? did the python community develop a canonical style for this?

To my eyes `foo: int=5` looks ok. Also, PEP 8 wants no spaces around the equal sign for default parameters, so I used none here.

File: doc/libalsaaudio.rst [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
 The :mod:`alsaaudio` module defines functions and classes for using ALSA.
 
-.. function:: pcms(pcmtype=PCM_PLAYBACK)
+.. function:: pcms(pcmtype:int=PCM_PLAYBACK) ->list[str]
```
Comment (ossilator) [inline doc/libalsaaudio.rst:13]: ...

File: alsaaudio.pyi [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+MIXER_SCHN_SIDE_LEFT: int
+MIXER_SCHN_SIDE_RIGHT: int
+MIXER_SCHN_REAR_CENTER: int
+MIXER_SCHN_MONO: int
+
+VOLUME_UNITS_PERCENTAGE: int
+VOLUME_UNITS_RAW: int
+VOLUME_UNITS_DB: int
+
+def pcms(pcmtype:int) -> list[str]: ...
```
Comment (ossilator) [inline alsaaudio.pyi:78]: missing space

File: alsaaudio.pyi [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	def getchannels() -> list[int]: ...
+	def setchannels(nchannels: int) -> None: ...
+	def pcmtype() -> int: ...
+	def pcmmode() -> int: ...
+	def cardname() -> str: ...
+	def setrate(rate: int) -> None: ...
+	def setformat(format: int) -> int: ...
+	def setperiodsize(period: int) -> int: ...
+	def read() -> tuple[int, bytes]: ...
+	def write(data:bytes) -> int: ...
```
Comment (ossilator) [inline alsaaudio.pyi:108]: missing space

File: alsaaudio.pyi [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	def avail() -> int: ...
+	def pause(enable:bool=True) -> int: ...
```
Comment (ossilator) [inline alsaaudio.pyi:110]: ...

File: doc/libalsaaudio.rst [hunk 9]
```diff
@@ -625,13 +625,13 @@ Mixer objects have the following methods:
    This method will return an empty tuple if the mixer is not an  enumerated
    control.
 
-.. method:: Mixer.setenum(index)
+.. method:: Mixer.setenum(index:int) -> None
```
Comment (ossilator) [inline doc/libalsaaudio.rst:628]: ...

**Review by ossilator — COMMENTED** [review]
you seem to still have missed some in the .pyi file. but did you touch the .rst file at all? there are still plenty; i just stopped marking them at some point.
you can easily locate them by regex-grepping for `:[^ ]`.

note also the still open request about close().

**Comment by larsimmisch** [comment]
Hi @ossilator,

I went through the PR again and fixed all whitespace issues. And a few missing type hints. I also undid the move of `close` (see https://github.com/larsimmisch/pyalsaaudio/pull/143#discussion_r1496436396). Hope this is complete/consistent now.

File: doc/libalsaaudio.rst [hunk 10]
```diff
@@ -675,14 +677,14 @@ Mixer objects have the following methods:
...
-.. method:: Mixer.getmute()
+.. method:: Mixer.getmute() -> list[int]
 
    Return a list indicating the current mute setting for each channel.
    0 means not muted, 1 means muted.
 
    This method will fail if the mixer has no playback switch capabilities.
 
-.. method:: Mixer.setmute(mute, [channel])
+.. method:: Mixer.setmute(mute: bool, channel: int|None=None) -> None
```
Comment (ossilator) [inline doc/libalsaaudio.rst:687]: missing spaces around `|`.

File: doc/libalsaaudio.rst [hunk 11]
```diff
@@ -692,14 +694,14 @@ Mixer objects have the following methods:
...
-.. method:: Mixer.getrec()
+.. method:: Mixer.getrec() -> list[int]
 
    Return a list indicating the current record mute setting for each channel.
    0 means not recording, 1 means recording.
 
    This method will fail if the mixer has no capture switch capabilities.
 
-.. method:: Mixer.setrec(capture, [channel])
+.. method:: Mixer.setrec(capture: int, channel: int|None=None) -> None
```
Comment (ossilator) [inline doc/libalsaaudio.rst:704]: missing spaces around `|`.

**Review by ossilator — CHANGES_REQUESTED** [review]
sorry, still a few issues left. :-(

given how easy i find it to spot these issues, i wonder why you don't. maybe you have an inappropriate font or font size configured? or is it just a perception difference? i find these breaks in the pattern jarring, so i just can't overlook them.

File: doc/libalsaaudio.rst [hunk 12]
```diff
@@ -658,7 +660,7 @@ Mixer objects have the following methods:
    The optional *units* argument can be one of :const:`VOLUME_UNITS_PERCENTAGE`,
    :const:`VOLUME_UNITS_RAW`, or :const:`VOLUME_UNITS_DB`.
 
-.. method:: Mixer.setvolume(volume, channel=None, pcmtype=PCM_PLAYBACK, units=VOLUME_UNITS_PERCENTAGE)
+.. method:: Mixer.setvolume(volume: int, pcmtype: int=PCM_PLAYBACK, units: int=VOLUME_UNITS_PERCENTAGE, channel: int|None) -> None
```
[the remark this answers is not in the document]
Comment (larsimmisch) [reply doc/libalsaaudio.rst:663]: That was deliberate (except for one inconsistency).

But I now see that https://docs.python.org/3/library/typing.html mostly has spaces before and after `|` (that document also uses `|` without surrounding spaces, but just in one case).

This reads ok to me with the exception of, say, `channel: int | None = None`. In these cases, I'll use `channel: (int | None) = None`.

File: doc/libalsaaudio.rst [hunk 13]
```diff
@@ -599,7 +601,7 @@ Mixer objects have the following methods:
    'Joined Capture Volume'   Manipulate sound capture volume for all channels at a time
    ========================  ================
 
-.. method:: Mixer.getenum()
+.. method:: Mixer.getenum() -> tuple[ str, list[str]]
```
[the remark this answers is not in the document]
Comment (larsimmisch) [reply doc/libalsaaudio.rst:604]: Fixed, thanks.

File: doc/libalsaaudio.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (ossilator) [reply doc/libalsaaudio.rst]: is that legal? i didn't read the document, but a search for `|` revealed no precedent for using parentheses.

File: doc/libalsaaudio.rst [hunk 14]
```diff
@@ -523,7 +525,7 @@ Mixer Objects
 
 Mixer objects provides access to the ALSA mixer API.
 
-.. class:: Mixer(control='Master', id=0, cardindex=-1, device='default')
+.. class:: Mixer(control: str = 'Master', id: int = 0, cardindex: int = -1, device str = 'default') -> Mixer
```
Comment (ossilator) [inline doc/libalsaaudio.rst:528]: missing colon after 'device'.

**Review by ossilator — CHANGES_REQUESTED** [review]

File: doc/libalsaaudio.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (larsimmisch) [reply doc/libalsaaudio.rst]: PyLance does not complain, but does complain when obvious syntax errors are present (like mismatched brackets).

[the remark this answers is not in the document]
Comment (ossilator) [reply doc/libalsaaudio.rst]: a more convincing test is just running it through real python. seems to work.

**Review by ossilator — APPROVED** [review]

**Comment by larsimmisch** [comment]
@ossilator thanks for the reviews
